wireapp/wire-ios-mono
# fix: sending message large group - WPB-11019

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-11019" title="WPB-11019" target="_blank"><img alt="Bug" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10803?size=medium" />WPB-11019</a>  [iOS]: In large group chats messages take long time to deliver
  </td></table>
  <br />
 





























































### Issue

On big groups, sending a message takes a lot of time with corecrypto via Proteus resulting quite often in the message not getting sent (expired).

This is due to two reasons: 
1. setting the expiration of the message is set before the encryption of the message so in big groups where we have a lot clients to encrypt a message for it can reach the timeout of 30 seconds.
2. For each encrypt data, we acquire and release the lock on corecrypto database to protect from other processes (NSE or Share Extension). During this process we also `restoreFromDisk` the information of corecrypto.

### Solution

1. set expiration date just before the message is sent via the API (after the encryption of message payload)
2. reduce the number of calls to encrypt the data for each clients of each users (of each domain) to a single call where we encrypt all necessary data for all proteusSessions (aka all clients at once)

To achieve this second point:
* extracted the information from CoreData models with a `MessageInfoExtractor`
* encrypt the payload and wrap it in protobuf objects with `ProteusMessagePayloadBuilder`
* pass the payload to the MessageAPI


### TODO

* [x] testing sending message | availability
* [ ] write tests for new objects

### Notes

I left the old encryptForTransport methods in order not to break legacy methods (WireCryptoBox) that could be used in tests - this will be cleaned in another PR.


### Testing

1. Send message in big groups (>100 users)
2. Check time spent looking for " "send message - attemptToSend duration:" logs

---

### Checklist


---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
